Describe `vshard.storage.on_bucket_event()` (#3310)
tarantool/doc: doc/reference/reference_rock/vshard/vshard_storage.rst

            ck_constraint: []
        ...

..  _storage_api-on_bucket_event:

..  function:: vshard.storage.on_bucket_event([trigger-function[, old-trigger-function]])

    Since vshard v.0.1.22. Define a trigger for execution when the data from
    the user spaces is changed (deleted or inserted) due to the rebalancing
    process. The trigger is invoked each time the data batch changes.

    :param function trigger-function: function which will become the trigger function.
    :param function old-trigger-function: existing trigger function which will
                                          be replaced by trigger-function.

    :return: nil or function pointer

    The ``trigger-function`` can have up to three parameters:

      * ``event_type`` (string) -- in order to distinguish event, you can compare
        this argument with the supported event types, ``bucket_data_recv_txn``
        and ``bucket_data_gc_txn``.

>       * ``bucket_id`` (unsigned) -- bucket id;
>       * ``data`` (table) -- additional information about data change transaction.
>         Currently it only includes an array of all spaces (``data.spaces``),

Reviewer: ```suggestion
      * ``bucket_id`` (unsigned) -- bucket id.
      * ``data`` (table) -- additional information about data change transaction.
        Currently, it only includes an array of all spaces (``data.spaces``),
```

Author: Rebased and fixed

            ck_constraint: []
        ...

..  _storage_api-on_bucket_event:

..  function:: vshard.storage.on_bucket_event([trigger-function[, old-trigger-function]])

    Since vshard v.0.1.22. Define a trigger for execution when the data from
    the user spaces is changed (deleted or inserted) due to the rebalancing

>     process. It's invoked on every data batch changed.

Reviewer: ```suggestion
    process. The trigger is invoked each time the data batch changes.
```

Author: Applied

>     :param trigger-function: function which will become the trigger function.
>     :param old-trigger-function: existing trigger function which will
>                                  be replaced by trigger-function.
> 
>     :return: nil or function pointer
> 
>     The trigger-function can have up to three parameters:
> 
>       * (string) event_type. In order to distinguish event you could
>         compare this argument with supported event types: ``bucket_data_recv_txn``
>         and ``bucket_data_gc_txn``;
>       * (unsigned) bucket_id;
>       * (table) data. E.g. includes an array of all spaces (data.spaces),
>         affected by a transaction, in which trigger-function is executed.

Reviewer: ```suggestion
    :param function trigger-function: function which will become the trigger function.
    :param function old-trigger-function: existing trigger function which will
                                 be replaced by trigger-function.

    :return: nil or function pointer
    
        The ``trigger-function`` can have up to three parameters:

      * ``event_type``(string) -- in order to distinguish event, you can
         compare this argument with the supported event types, ``bucket_data_recv_txn``
         and ``bucket_data_gc_txn``.
      * ``bucket_id`` (unsigned) -- bucket id.
      * ``data`` (table) -- it may include an array of all spaces (``data.spaces``)
        affected by a transaction in which trigger-function is executed.
```

Author: Applied with slight difference: `data` description seems to be too vague. Let's just update it, when new info appears in `data` and state for now, what is really inside it.

Author: The rendering is strange if spaces are added in 328 line. I'll omit it, if you don't mind

    **Example:**

    ..  code-block:: lua

        vshard.storage.on_bucket_event(function(event, bucket_id, data)
            if event == 'bucket_data_recv_txn' then
                -- Handle it.
                for idx, space in ipairs(data.spaces) do
                    ...
                end
            elseif event == 'bucket_data_gc_txn' then
                -- Handle it.
                ...
            end
        end)

>     .. NOTE::
> 
>         As everything executed inside triggers is already in a transaction,
>         you shouldn't use transactions, yield-operations, changes to different
>         space engines.
> 
>         If the parameters are ``(nil, old-trigger-function)``,
>         then the old trigger is deleted.
> 
>         If both parameters are omitted, then the response is a list of
>         existing trigger functions.

Reviewer: ```suggestion
    ..  NOTE::

        As everything executed inside triggers is already in a transaction,
        you shouldn't use transactions, yield-operations
        (:ref:`explicit <app-yields>` or not), or changes 
        to different space engines (see :ref:`rule #2 <box-txn_management>`).

    If the parameters are ``(nil, old-trigger-function)``,
    then the old trigger is deleted.

    If both parameters are omitted, then the response is a list of
    existing trigger functions.
```

Author: This one is pretty default note for triggers, e.g.: https://www.tarantool.io/en/doc/latest/reference/reference_lua/net_box/#net-box-on-schema-reload.

I'd rather let all info remain inside note, if you don't mind. If you insist, I'll apply. 

Links were applied
